clients/src/main/java/org/apache/kafka/common/requests/ProduceResponse.java

```diff
@@ -225,13 +225,13 @@ public ProduceResponse(Struct struct) {
                 long logAppendTime = partRespStruct.getLong(LOG_APPEND_TIME_KEY_NAME);
                 long logStartOffset = partRespStruct.getOrElse(LOG_START_OFFSET_FIELD, INVALID_OFFSET);
 
-                Map<Integer, String> errorRecords = new HashMap<>();
+                List<ErrorRecord> errorRecords = new ArrayList<>();
```

Comment:
Record errors are unlikely, so maybe we can initialize this data structure lazily. Maybe something as simple as this:
```java
List<ErrorRecord> errorRecords = Collections.emptyList();
if (partRespStruct.hasField(ERROR_RECORDS_KEY_NAME)) {
  Object[] recordErrorsArray = partRespStruct.getArray(ERROR_RECORDS_KEY_NAME)
  if (recordErrorsArray.length > 0) {
    errorRecords = new ArrayList<>(recordErrorsArray.length);
    ...
  }
}
```